### PR TITLE
Add support for NIOTransportServices

### PR DESCRIPTION
**Pull request by glbrntt**

### CODE REVIEW

File: Sources/GRPC/GRPCNIO.swift [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      return implementation
+    }
+  }
+}
+
+// MARK: - Generic Bootstraps
+
+/// This protocol is intended as a layer of abstraction over `ClientBootstrap` and
+/// `NIOTSConnectionBootstrap`.
+public protocol GenericClientBootstrap {
```
Comment (MrMage) [inline Sources/GRPC/GRPCNIO.swift]: How about e.g. `TransportAgnosticClientBootstrap`? Just an idea; not sure if that's any better.

Also, couldn't this be part of NIO, given its universal usefulness?

File: Sources/GRPC/GRPCNIO.swift [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+@available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *)
+extension NIOTSListenerBootstrap: GenericServerBootstrap {}
+#endif
+
+// MARK: - Bootstrap / EventLoopGroup helpers
+
+public enum GRPCNIO {
+  /// Makes a new event loop group based on the network preference.
+  ///
+  /// If `.best` is chosen and `Network.framework` is available then ` NIOTSEventLoopGroup` will
```
Comment (MrMage) [inline Sources/GRPC/GRPCNIO.swift]: super-nit: remove space in ` NIOTSEventLoopGroup`

File: Sources/GRPC/GRPCNIO.swift (hunk 1 above)
Comment (glbrntt) [reply Sources/GRPC/GRPCNIO.swift]: Hmm I'm not sure about the naming. How about `ClientBootstrapProtocol`? Swift has `StringProtocol` and `IteratorProtocol` so I don't think that it's too much of a stretch.

Comment (glbrntt) [reply Sources/GRPC/GRPCNIO.swift]: Re: NIO: https://github.com/apple/swift-nio/issues/796

Comment (MrMage) [reply Sources/GRPC/GRPCNIO.swift]: Sounds good!

Comment (MrMage) [reply Sources/GRPC/GRPCNIO.swift]: But should we add a TODO to remove this once https://github.com/apple/swift-nio/issues/796 is fixed?

Comment (MrMage) [reply Sources/GRPC/GRPCNIO.swift]: But should we add a TODO to remove this once https://github.com/apple/swift-nio/issues/796 is fixed?

**Comment by MrMage** [comment]
Please let me know if/when this is ready to merge :-)

**Comment by glbrntt** [comment]
@MrMage ready to merge

**Pull request merged**